src/Microsoft.ComponentDetection.Common/DependencyGraph/ComponentRecorder.cs

```diff
@@ -85,25 +81,26 @@ public ISingleFileComponentRecorder CreateSingleFileComponentRecorder(string loc
             throw new ArgumentNullException(nameof(location));
         }
 
-        var matching = this.singleFileRecorders.FirstOrDefault(x => x.ManifestFileLocation == location);
-        if (matching == null)
-        {
-            matching = new SingleFileComponentRecorder(location, this, this.enableManualTrackingOfExplicitReferences, this.logger);
-            this.singleFileRecorders.Add(matching);
-        }
-
-        return matching;
+        return this.singleFileRecorders.GetOrAdd(location, loc => new SingleFileComponentRecorder(loc, this, this.enableManualTrackingOfExplicitReferences, this.logger));
     }
 
     public IReadOnlyDictionary<string, IDependencyGraph> GetDependencyGraphsByLocation()
     {
-        return this.singleFileRecorders.Where(x => x.DependencyGraph.HasComponents())
-            .ToImmutableDictionary(x => x.ManifestFileLocation, x => x.DependencyGraph as IDependencyGraph);
+        return new ReadOnlyDictionary<string, IDependencyGraph>(
+            this.singleFileRecorders.Values
+                .Where(x => x.DependencyGraph.HasComponents())
+                .ToDictionary(x => x.ManifestFileLocation, x => x.DependencyGraph as IDependencyGraph));
```

Comment:
Just out of curiosity what is the benefit of this change?

Reply:
`Dictionary` implements `IReadOnlyDictionary` which is what the consumer expects and is more performant than `ImmutableDictionary`.

In general, previously this class uses `ConcurrentBag`, but the way it is used is just like a dictionary from location to recorder. Some computation like `GetDependencyGraphForLocation` can be better because of O(1) lookup instead of searching for location linearly.